solver/llbsolver/ops/exec.go

```diff
@@ -90,20 +90,52 @@ func cloneExecOp(old *pb.ExecOp) pb.ExecOp {
 	n.Mounts = nil
 	for i := range old.Mounts {
 		m := *old.Mounts[i]
+
+		if m.CacheOpt != nil {
+			co := *m.CacheOpt
+			m.CacheOpt = &co
+		}
+
 		n.Mounts = append(n.Mounts, &m)
 	}
 	return n
 }
 
 func (e *ExecOp) CacheMap(ctx context.Context, g session.Group, index int) (*solver.CacheMap, bool, error) {
 	op := cloneExecOp(e.op)
+
 	for i := range op.Meta.ExtraHosts {
 		h := op.Meta.ExtraHosts[i]
 		h.IP = ""
 		op.Meta.ExtraHosts[i] = h
 	}
+
+	checkShouldClearCacheOpts := func(m *pb.Mount) bool {
```

Comment:
This is included per slack discussion with @tonistiigi 

I'm not sure its really necessary as the only thing I can think this would effect is that without it existing caches will miss due to the ID no longer being accounted for in the cache map.

The intention here was that the dockerfile frontend sets `m.CacheOpt.ID` to `m.Dest` unless otherwise specified.

Reply:
Move this to a separate helper function instead of inlining. Reference: https://github.com/moby/buildkit/blob/master/solver/llbsolver/ops/file.go#L634-L637

Reply:
Done.